<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="model" default="waffle_pi_v5_complex" doc="model type [burger, waffle, waffle_pi]"/>
    <arg name="first_tb3"  default="amigobot_1"/>
    <arg name="second_tb3" default="amigobot_2"/>
    <arg name="third_tb3"  default="amigobot_3"/>
    <arg name="fourth_tb3" default="amigobot_4"/>
    <arg name="fiveth_tb3" default="amigobot_5"/>

    <arg name="first_tb3_x_pos" default=" -2.15"/>
    <arg name="first_tb3_y_pos" default="  0.0"/>
    <arg name="first_tb3_z_pos" default="  0.0"/>
    <arg name="first_tb3_yaw"   default="  0.0"/>

    <arg name="second_tb3_x_pos" default=" 2.15"/>
    <arg name="second_tb3_y_pos" default=" 0.0"/>
    <arg name="second_tb3_z_pos" default=" 0.0"/>
    <arg name="second_tb3_yaw"   default="-3.14"/>

    <!--11-->
    <arg name="third_tb3_x_pos" default="  5.15"/>
    <arg name="third_tb3_y_pos" default="  7.0"/>
    <arg name="third_tb3_z_pos" default="  0.0"/>
    <arg name="third_tb3_yaw"   default="  1.57"/>

    <!--26-->
    <arg name="fourth_tb3_x_pos" default="-10.65"/>
    <arg name="fourth_tb3_y_pos" default=" 2.5"/>
    <arg name="fourth_tb3_z_pos" default=" 0.0"/>
    <arg name="fourth_tb3_yaw"   default="-1.57"/>

    <!--23-->
    <arg name="fiveth_tb3_x_pos" default="-4.9"/>
    <arg name="fiveth_tb3_y_pos" default=" 1.25"/>
    <arg name="fiveth_tb3_z_pos" default=" 0.0"/>
    <arg name="fiveth_tb3_yaw"   default="-3.14"/>

    <include file="$(find gazebo_ros)/launch/empty_world.launch">
      <!--arg name="world_name" value="$(find turtlebot3_gazebo)/worlds/empty_10x.world"/-->
      <arg name="world_name" value="$(find amigobot_LTL)/gazebo/empty.world"/>
      <arg name="paused" value="false"/>
      <arg name="use_sim_time" value="true"/>
      <arg name="gui" value="true"/>
      <arg name="headless" value="false"/>
      <arg name="debug" value="false"/>
    </include>  

    <group ns = "$(arg first_tb3)">
      <param name="robot_description" command="$(find xacro)/xacro --inorder $(find amigobot_LTL)/model/urdf/turtlebot3_$(arg model).urdf.xacro" />

      <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
        <param name="publish_frequency" type="double" value="50.0" />
        <param name="tf_prefix" value="$(arg first_tb3)" />
      </node>
      
      <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg first_tb3) -x $(arg first_tb3_x_pos) -y $(arg first_tb3_y_pos) -z $(arg first_tb3_z_pos) -Y $(arg first_tb3_yaw) -param robot_description" />
    </group>

    <group ns = "$(arg second_tb3)">
      <param name="robot_description" command="$(find xacro)/xacro --inorder $(find amigobot_LTL)/model/urdf/turtlebot3_$(arg model).urdf.xacro" />

      <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
        <param name="publish_frequency" type="double" value="50.0" />
        <param name="tf_prefix" value="$(arg second_tb3)" />
      </node>

      <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg second_tb3) -x $(arg second_tb3_x_pos) -y $(arg second_tb3_y_pos) -z $(arg second_tb3_z_pos) -Y $(arg second_tb3_yaw) -param robot_description" />
    </group>

    <group ns = "$(arg third_tb3)">
      <param name="robot_description" command="$(find xacro)/xacro --inorder $(find amigobot_LTL)/model/urdf/turtlebot3_$(arg model).urdf.xacro" />

      <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
        <param name="publish_frequency" type="double" value="50.0" />
        <param name="tf_prefix" value="$(arg third_tb3)" />
      </node>

      <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg third_tb3) -x $(arg third_tb3_x_pos) -y $(arg third_tb3_y_pos) -z $(arg third_tb3_z_pos) -Y $(arg third_tb3_yaw) -param robot_description" />
    </group>

    <group ns = "$(arg fourth_tb3)">
      <param name="robot_description" command="$(find xacro)/xacro --inorder $(find amigobot_LTL)/model/urdf/turtlebot3_$(arg model).urdf.xacro" />

      <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
        <param name="publish_frequency" type="double" value="50.0" />
        <param name="tf_prefix" value="$(arg fourth_tb3)" />
      </node>

      <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg fourth_tb3) -x $(arg fourth_tb3_x_pos) -y $(arg fourth_tb3_y_pos) -z $(arg fourth_tb3_z_pos) -Y $(arg fourth_tb3_yaw) -param robot_description" />
    </group>

    <group ns = "$(arg fiveth_tb3)">
      <param name="robot_description" command="$(find xacro)/xacro --inorder $(find amigobot_LTL)/model/urdf/turtlebot3_$(arg model).urdf.xacro" />

      <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
        <param name="publish_frequency" type="double" value="50.0" />
        <param name="tf_prefix" value="$(arg fiveth_tb3)" />
      </node>

      <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg fiveth_tb3) -x $(arg fiveth_tb3_x_pos) -y $(arg fiveth_tb3_y_pos) -z $(arg fiveth_tb3_z_pos) -Y $(arg fiveth_tb3_yaw) -param robot_description" />
    </group>

    <node pkg="amigobot_LTL" name="amigobot_LTL" type="main_complex_v5_original.py" output="screen">
    </node>

        <node pkg="rviz" name="rviz" type="rviz" output="screen" args="-d /home/ubuntu484/catkin_ws/src/amigobot_LTL/launch/complex/rviz/complex_v5.rviz">
    </node>
    
    <!-- for showing path in rviz-->
    <node pkg="tf" type="static_transform_publisher" name="$(arg first_tb3)_map_odom_broadcaster"  args="0 0 0 0 0 0 /map /$(arg first_tb3)/odom 5" />
    <node pkg="tf" type="static_transform_publisher" name="$(arg second_tb3)_map_odom_broadcaster" args="0 0 0 0 0 0 /map /$(arg second_tb3)/odom 5" />
    <node pkg="tf" type="static_transform_publisher" name="$(arg third_tb3)_map_odom_broadcaster"  args="0 0 0 0 0 0 /map /$(arg third_tb3)/odom 5" />    
    <node pkg="tf" type="static_transform_publisher" name="$(arg fourth_tb3)_map_odom_broadcaster" args="0 0 0 0 0 0 /map /$(arg fourth_tb3)/odom 5" />
    <node pkg="tf" type="static_transform_publisher" name="$(arg fiveth_tb3)_map_odom_broadcaster" args="0 0 0 0 0 0 /map /$(arg fiveth_tb3)/odom 5" />

</launch>
